diff -ruNp gnome-shell-3.2.2.1.orig/src/gnome-shell-plugin.c gnome-shell-3.2.2.1/src/gnome-shell-plugin.c
--- gnome-shell-3.2.2.1.orig/src/gnome-shell-plugin.c	2012-01-19 23:16:22.000000000 -0600
+++ gnome-shell-3.2.2.1/src/gnome-shell-plugin.c	2012-05-10 09:15:04.752169698 -0500
@@ -30,8 +30,6 @@
 
 #include <clutter/clutter.h>
 #include <clutter/x11/clutter-x11.h>
-#include <GL/glx.h>
-#include <GL/glxext.h>
 #include <gjs/gjs.h>
 #include <meta/display.h>
 #include <meta/meta-plugin.h>
@@ -146,14 +144,18 @@ static void
 gnome_shell_plugin_start (MetaPlugin *plugin)
 {
   GnomeShellPlugin *shell_plugin = GNOME_SHELL_PLUGIN (plugin);
+#if 0
   MetaScreen *screen;
   MetaDisplay *display;
   Display *xdisplay;
+#endif
   GError *error = NULL;
   int status;
+#if 0  
   const char *glx_extensions;
+#endif
   GjsContext *gjs_context;
-
+#if 0
   screen = meta_plugin_get_screen (plugin);
   display = meta_screen_get_display (screen);
 
@@ -166,7 +168,8 @@ gnome_shell_plugin_start (MetaPlugin *pl
   glx_extensions = glXQueryExtensionsString (xdisplay,
                                              meta_screen_get_screen_number (screen));
   shell_plugin->have_swap_event = strstr (glx_extensions, "GLX_INTEL_swap_event") != NULL;
-
+#endif
+  shell_plugin->have_swap_event = 0;
   shell_perf_log_define_event (shell_perf_log_get_default (),
                                "glx.swapComplete",
                                "GL buffer swap complete event received (with timestamp of completion)",
