--- ./gst-libs/ext/ffmpeg/configure.orig	2010-10-11 03:51:49.677970647 +0200
+++ ./gst-libs/ext/ffmpeg/configure	2010-10-11 03:52:22.576680789 +0200
@@ -2731,8 +2731,6 @@
     add_cflags -msg_disable nomainieee
     add_cflags -msg_disable ptrmismatch1
     add_cflags -msg_disable unreachcode
-elif enabled gcc; then
-    check_cflags -fno-tree-vectorize
 elif enabled clang; then
     check_cflags -Qunused-arguments
 elif enabled armcc; then
